### PR TITLE
image-sets: implement image-sets/view/{imageSetID} ... details endpoints.

### PR DESCRIPTION
**Pull request by ldjebran**
Fixes #THEEDGE-2317
## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [X] Refactor

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/routes/imagesets.go [hunk 1]
```diff
@@ -402,3 +413,188 @@ func GetImageSetsView(w http.ResponseWriter, r *http.Request) {
...
+		imageID, err := strconv.Atoi(imageIDString)
+		if err != nil {
+			ctxServices.Log.WithField("error", err.Error()).Error("error while converting image id from string")
+			respondWithAPIError(w, ctxServices.Log, errors.NewBadRequest("bad image id"))
+			return
+		}
+		var image models.Image
+		if result := db.Org(orgID, "").Where("image_set_id", imageSet.ID).First(&image, imageID); result.Error != nil {
+			if result.Error == gorm.ErrRecordNotFound {
+				respondWithAPIError(w, ctxServices.Log, errors.NewNotFound("image not found"))
```
Comment (AaronH88) [inline pkg/routes/imagesets.go:545]: Can we make this error a little more explicit so its easier to debug?

e.g. "image set for image set view not found"

File: pkg/routes/imagesets.go [hunk 2]
```diff
@@ -402,3 +413,188 @@ func GetImageSetsView(w http.ResponseWriter, r *http.Request) {
...
+				return
+			}
+			apiError := errors.NewInternalServerError()
+			apiError.SetTitle("internal server error occurred while getting image")
```
Comment (AaronH88) [inline pkg/routes/imagesets.go:549]: "while getting image set"

File: pkg/services/images.go [hunk 3]
```diff
@@ -1518,3 +1520,68 @@ func (s *ImageService) SetDevicesUpdateAvailabilityFromImageSet(orgID string, Im
...
+	if err != nil {
+		return 0, err
+	}
+
+	if tx == nil {
+		tx = db.DB
+	}
+
+	var count int64
+	result := db.OrgDB(orgID, tx, "").Debug().Model(&models.Image{}).Count(&count)
```
Comment (AaronH88) [inline pkg/services/images.go:1536]: why `.Debug()` ?

File: pkg/routes/imagesets.go (hunk 2 above)
Comment (ldjebran) [reply pkg/routes/imagesets.go:549]: here we are getting image,
so maybe 
 "while getting image set image"

File: pkg/services/imagesets.go [hunk 4]
```diff
@@ -146,32 +207,160 @@ func (s *ImageSetsService) GetImageSetsView(limit int, offset int, tx *gorm.DB)
...
-		installerID, ok := imageSetsInstallersMap[imageSetRow.ID]
-		if ok {
-			imageSetView.ImageBuildIsoURL = GetStorageInstallerIsoURL(installerID)
+		s.log.WithFields(log.Fields{"error": result.Error.Error(), "OrgID": orgID}).Error("error getting image-set")
+		return nil, result.Error
+	}
+
+	var lastImage models.Image
+	// get the last image-set image
+	if result := db.Org(orgID, "").Debug().Order("created_at DESC").
```
Comment (AaronH88) [inline pkg/services/imagesets.go:265]: Should `Debug()` be here?

I see why you ordered by `created_at` , should we be ordering by `version` instead? I dont think it makes a difference to the functionality so im ok to keep it this way, version just makes more sense in my head :smile:

File: pkg/services/imagesets.go [hunk 5]
```diff
@@ -146,32 +207,160 @@ func (s *ImageSetsService) GetImageSetsView(limit int, offset int, tx *gorm.DB)
...
+		if result.Error == gorm.ErrRecordNotFound {
+			s.log.WithFields(log.Fields{"error": result.Error.Error(), "OrgID": orgID}).Error("image-set last image not found")
+			return nil, new(ImageNotFoundError)
 		}
-		imageSetsView = append(imageSetsView, imageSetView)
+		s.log.WithFields(log.Fields{"error": result.Error.Error(), "OrgID": orgID}).Error("error getting image-set last-image")
+		return nil, result.Error
+	}
+
+	imageInfo, err := imageService.AddPackageInfo(&lastImage)
```
Comment (AaronH88) [inline pkg/services/imagesets.go:281]: Does the `Preload("Packages")` on line 266 not add this package info? 

Should we remove this preload from the DB query if we are going to call this function regardless?

File: pkg/services/imagesets.go [hunk 6]
```diff
@@ -146,32 +207,160 @@ func (s *ImageSetsService) GetImageSetsView(limit int, offset int, tx *gorm.DB)
...
+		if result.Error == gorm.ErrRecordNotFound {
+			s.log.WithFields(log.Fields{"error": result.Error.Error(), "OrgID": orgID}).Error("image-set not found")
+			return nil, new(ImageSetNotFoundError)
+		}
+		return nil, result.Error
+	}
+
+	var image models.Image
+	// get the image-set image
+	if result := db.Org(orgID, "").Debug().Order("created_at DESC").
```
Comment (AaronH88) [inline pkg/services/imagesets.go:331]: Same comments as in the `GetImageSetViewByID` func :smile:

File: pkg/services/imagesets.go (hunk 4 above)
Comment (ldjebran) [reply pkg/services/imagesets.go:265]: sorting by created_at at this I am sure that I will have the latest first , but with version if value is missed or missed, this portion of code will have some troubles, that's why relaying on created_at is more robust here

File: pkg/services/imagesets.go (hunk 5 above)
Comment (ldjebran) [reply pkg/services/imagesets.go:281]: this function call is for one image, not like the old one that was calling this for all images.  AddPackageInfo  is different from Preload("Packages")  
AddPackageInfo will return the difference packages from last image 
```go
	imgDetail.UpdateAdded = len(upd[len(upd)-1].PackageDiff.Removed)
		imgDetail.UpdateRemoved = len(upd[len(upd)-1].PackageDiff.Added)
		imgDetail.UpdateUpdated = len(upd[len(upd)-1].PackageDiff.Upgraded)
```

File: pkg/services/imagesets.go (hunk 6 above)
Comment (ldjebran) [reply pkg/services/imagesets.go:331]: same response :smile:

File: pkg/services/imagesets.go (hunk 4 above)
Comment (ldjebran) [reply pkg/services/imagesets.go:265]: we are also able to sort by ID DESC , but do not like to sort by ids :smile:

File: pkg/services/images.go (hunk 3 above)
Comment (ldjebran) [reply pkg/services/images.go]: removed all debug :smile:

File: pkg/services/imagesets.go (hunk 4 above)
Comment (ldjebran) [reply pkg/services/imagesets.go]: removed the debug

Comment (AaronH88) [reply pkg/services/imagesets.go]: Sounds good to me, ty

**Review by AaronH88 — APPROVED** [review]
/lgtm

**Pull request merged**